File: Claude-decides/made/dla.py
"""
Diffusion-Limited Aggregation (DLA)

A particle is released at a random point on the boundary.
It walks randomly — Brownian motion — until it either:
  - Touches the growing cluster: it sticks and becomes part of the cluster
  - Wanders too far from the cluster: it is released and a new one starts

Starting from a single seed particle at the center,
this process produces branching, fractal structures:
coral, lightning, river deltas, frost on glass, mineral dendrites.

Why branches?

The tips of the cluster reach out into empty space and are
more likely to be hit by incoming particles than the interior
(which is shadowed by the tips). The tips grow; the interior doesn't.
This feedback — tips attract tips — produces branching at every scale.

The resulting structure is a fractal with dimension ≈ 1.71
(between a line and a plane — denser than a line, sparser than a solid).

Real-world DLA and DLA-like processes:
  - Zinc crystals grown from solution
  - Lightning channels (dielectric breakdown)
  - River delta formation
  - Snowflake dendrite growth
  - Bacterial colony growth under certain conditions

The branching is not programmed in. It emerges from:
  random walks + aggregation + geometric shadowing.
"""

import random
import math

W, H = 71, 35
CX, CY = W // 2, H // 2
MAX_PARTICLES = 800
ASPECT = 2.0  # chars are ~2x taller than wide

# Radius at which we release new particles (and abandon lost ones)
R_RELEASE_FACTOR = 1.1
R_KILL_FACTOR = 1.5

# Characters to show depth of attachment order
CHARS = '·:;+=xX$#@█'


def dist(x, y):
    # Aspect-correct distance (account for terminal char proportions)
    return math.sqrt(((x - CX) / 1.0) ** 2 + ((y - CY) / ASPECT) ** 2)


def neighbors(x, y):
    return [(x+1, y), (x-1, y), (x, y+1), (x, y-1)]


def run():
    grid = {}        # (x, y) -> particle number (order it was deposited)
    occupied = set() # just the coordinates for fast lookup

    # Seed particle at center
    occupied.add((CX, CY))
    grid[(CX, CY)] = 0

    r_cluster = 1.0   # current cluster radius

    for i in range(1, MAX_PARTICLES + 1):
        # Release radius: just outside the current cluster
        r_release = r_cluster * R_RELEASE_FACTOR + 2
        r_kill = r_cluster * R_KILL_FACTOR + 5

        # Random point on a circle at r_release
        theta = random.uniform(0, 2 * math.pi)
        px = int(CX + r_release * math.cos(theta))
        py = int(CY + r_release * math.sin(theta) * ASPECT)

        # Random walk until stick or die
        stuck = False
        for _ in range(100000):
            # Check if adjacent to cluster
            for nx, ny in neighbors(px, py):
                if (nx, ny) in occupied:
                    # Stick here
                    if 0 <= px < W and 0 <= py < H:
                        occupied.add((px, py))
                        grid[(px, py)] = i
                        d = dist(px, py)
                        if d > r_cluster:
                            r_cluster = d
                    stuck = True
                    break

            if stuck:
                break

            # Random step
            dx, dy = random.choice([(1,0),(-1,0),(0,1),(0,-1)])
            px += dx
            py += dy

            # Kill if too far
            if dist(px, py) > r_kill:
                break

    return grid


def render(grid, total):
    canvas = [[' '] * W for _ in range(H)]

    for (x, y), order in grid.items():
        if 0 <= x < W and 0 <= y < H:
            # Color by order: early (center) = dense chars, late (tips) = lighter
            progress = order / total
            idx = int(progress * (len(CHARS) - 1))
            canvas[y][x] = CHARS[idx]

    for row in canvas:
        print('  ' + ''.join(row))


def main():
    print('Diffusion-Limited Aggregation\n')
    print('  Particles random-walk until they touch the growing cluster.')
    print('  Tips grow fastest: they intercept walkers before the interior can.')
    print('  Result: branching at every scale — fractal dimension ≈ 1.71.\n')
    print('  · = first particles deposited (center)    █ = most recent (tips)\n')
    print('  Growing...')

    grid = run()

    print(f'  {len(grid)} particles deposited.\n')
    render(grid, MAX_PARTICLES)

    print()
    print('  Why the branches? The tips are exposed; the interior is shadowed.')
    print('  Particles reach the tips before they can diffuse to the interior.')
    print('  The shape amplifies itself — each tip makes the next tip more likely.')
    print()
    print('  Lightning works the same way: the leader tip is most exposed,')
    print('  so the next discharge follows it. The branch is the path.')
    print()
    print('  This is not lightning deciding to branch.')
    print('  Branching is the only thing that can happen here.')


if __name__ == '__main__':
    main()

File: site/public/scripts/dla.js
// Diffusion-Limited Aggregation — growing fractal cluster
// Particles random-walk from the boundary until they touch the cluster and stick.
// The branching geometry emerges from the physics of diffusion — tips grow faster
// because they intercept more random walkers.
const canvas = document.getElementById('program-canvas');
const ctx = canvas.getContext('2d');

let animId;

// Work at reduced resolution for performance, scale up to canvas
const SCALE = 2;
let GW, GH;
let grid;         // Uint8Array: 0=empty, 1=cluster
let clusterCount = 0;
let maxDist = 0;   // furthest particle from center (for release radius)

const TARGET = 2500;

let cx, cy;

function gridIdx(x, y) { return y * GW + x; }

function init() {
  cancelAnimationFrame(animId);
  const W = canvas.width, H = canvas.height;
  GW = Math.floor(W / SCALE);
  GH = Math.floor(H / SCALE);
  grid = new Uint8Array(GW * GH);
  clusterCount = 0;

  cx = Math.floor(GW / 2);
  cy = Math.floor(GH / 2);

  // Seed particle at center
  grid[gridIdx(cx, cy)] = 1;
  clusterCount = 1;
  maxDist = 5;

  // Draw background
  ctx.fillStyle = '#0f0f0f';
  ctx.fillRect(0, 0, W, H);

  // Draw seed
  drawParticle(cx, cy, 0);

  window.__setStatus && window.__setStatus('0 particles — click to restart');
  animId = requestAnimationFrame(frame);
}

// Color by normalized distance from center: warm amber → cool blue-white
function particleColor(dist, maxD) {
  const t = Math.min(1, dist / (maxD * 0.85 + 1));
  const r = Math.round(200 - t * 70);
  const g = Math.round(110 + t * 80);
  const b = Math.round(20 + t * 200);
  return `rgb(${r},${g},${b})`;
}

function drawParticle(gx, gy, dist) {
  ctx.fillStyle = particleColor(dist, maxDist);
  ctx.fillRect(gx * SCALE, gy * SCALE, SCALE, SCALE);
}

function dist2(x, y) {
  const dx = x - cx, dy = y - cy;
  return Math.sqrt(dx * dx + dy * dy);
}

// Try to add one particle; returns true if it stuck
function addOneParticle() {
  const releaseR = Math.min(maxDist + 8, Math.min(GW, GH) / 2 - 2);
  const killR = releaseR + 15;

  // Release from random point on a circle of radius releaseR
  const angle = Math.random() * 2 * Math.PI;
  let px = Math.round(cx + releaseR * Math.cos(angle));
  let py = Math.round(cy + releaseR * Math.sin(angle));

  // Clamp to grid
  px = Math.max(1, Math.min(GW - 2, px));
  py = Math.max(1, Math.min(GH - 2, py));

  const MAX_STEPS = 5000;
  for (let step = 0; step < MAX_STEPS; step++) {
    // Random walk step
    const dir = Math.floor(Math.random() * 4);
    if (dir === 0) px++;
    else if (dir === 1) px--;
    else if (dir === 2) py++;
    else py--;

    // Boundary wrap / kill
    if (px <= 0 || px >= GW - 1 || py <= 0 || py >= GH - 1) return false;
    const d = dist2(px, py);
    if (d > killR) return false;

    // Check 4 neighbors for cluster contact
    if (
      grid[gridIdx(px+1, py)] ||
      grid[gridIdx(px-1, py)] ||
      grid[gridIdx(px, py+1)] ||
      grid[gridIdx(px, py-1)]
    ) {
      // Stick!
      grid[gridIdx(px, py)] = 1;
      clusterCount++;
      const particleDist = d;
      if (particleDist > maxDist) maxDist = particleDist;
      drawParticle(px, py, particleDist);
      return true;
    }
  }
  return false;
}

const BATCH = 3;

function frame() {
  if (clusterCount >= TARGET) {
    window.__setStatus && window.__setStatus(`${clusterCount} particles — fractal complete — click to restart`);
    return;
  }

  for (let i = 0; i < BATCH; i++) {
    addOneParticle();
  }

  window.__setStatus && window.__setStatus(`${clusterCount} particles — click to restart`);
  animId = requestAnimationFrame(frame);
}

window.__programRestart = init;
init();
